abe/fame.go

```diff
@@ -0,0 +1,343 @@
+// determining which attributes are needed for an entity to be able
+// to decrypt. Moreover, secret keys are generated, where each key
+// is connected to some attribute, such that only a set of keys whose
+// attributes are sufficient can decrypt the massage.
+// This scheme is a PUBLIC-KEY scheme - no master secret key is needed
+// to encrypt the messages.
+//
+
+// FAME represents a FAME scheme.
+type FAME struct {
+	P *big.Int // order of the elliptic curve
+}
+
+// NewFAME configures a new instance of the scheme.
+func NewFAME() *FAME {
+	return &FAME{P: bn256.Order}
+}
+
+// FAMESecKey represents a master secret key of a FAME scheme.
+type FAMESecKey struct {
+	partInt [4]*big.Int
+	partG1  [3]*bn256.G1
+}
+
+// FAMEPubKey represents a public key of a FAME scheme.
+type FAMEPubKey struct {
+	partG2 [2]*bn256.G2
+	partGT [2]*bn256.GT
+}
+
+// GenerateMasterKeys generates a new set of public keys, needed
+// for encrypting data, and master secret keys needed for generating
+// keys for decrypting.
+func (a *FAME) GenerateMasterKeys() (*FAMEPubKey, *FAMESecKey, error) {
+	sampler := sample.NewUniformRange(big.NewInt(1), a.P)
+	val, err := data.NewRandomVector(7, sampler)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	partInt := [4]*big.Int{val[0], val[1], val[2], val[3]}
+	partG1 := [3]*bn256.G1{new(bn256.G1).ScalarBaseMult(val[4]),
+		new(bn256.G1).ScalarBaseMult(val[5]),
+		new(bn256.G1).ScalarBaseMult(val[6])}
+	partG2 := [2]*bn256.G2{new(bn256.G2).ScalarBaseMult(val[0]),
+		new(bn256.G2).ScalarBaseMult(val[1])}
+	tmp1 := new(big.Int).Mod(new(big.Int).Add(new(big.Int).Mul(val[0], val[4]), val[6]), a.P)
+	tmp2 := new(big.Int).Mod(new(big.Int).Add(new(big.Int).Mul(val[1], val[5]), val[6]), a.P)
+	partGT := [2]*bn256.GT{new(bn256.GT).ScalarBaseMult(tmp1),
+		new(bn256.GT).ScalarBaseMult(tmp2)}
+
+	return &FAMEPubKey{partG2: partG2, partGT: partGT},
+		&FAMESecKey{partInt: partInt, partG1: partG1}, nil
+}
+
+// FAMECipher represents a ciphertext of a FAME scheme.
+type FAMECipher struct {
+	ct0     [3]*bn256.G2
+	ct      [][3]*bn256.G1
+	ctPrime *bn256.GT
+	msp     *MSP
+}
+
+// Encrypt takes as an input a message msg represented as an element of an elliptic
+// curve, a MSP struct representing the decryption policy, and a public key pk. It
+// returns an encryption of the message. In case of a failed procedure an error
+// is returned. Note that safety of the encryption is only proved if the mapping
+// msp.RowToAttrib from the rows of msp.Mat to attributes is injective.
+func (a *FAME) Encrypt(msg *bn256.GT, msp *MSP, pk *FAMEPubKey) (*FAMECipher, error) {
+	if len(msp.Mat) == 0 || len(msp.Mat[0]) == 0 {
+		return nil, fmt.Errorf("empty msp matrix")
+	}
+
+	attrib := make(map[int]bool)
+	for _, i := range msp.RowToAttrib {
+		if attrib[i] {
+			return nil, fmt.Errorf("some attributes correspond to" +
+				"multiple rows of the MSP struct, the scheme is not secure")
+		} else {
+			attrib[i] = true
```

Comment:
Nitpick: the `else` isn't needed since `attrib[i] = true` will be executed every time (provided that no errors occur).